README.md
# panopticon-tui

![CI](https://github.com/ScalaConsultants/panopticon-tui/workflows/Rust%20CI/badge.svg)
![crates.io](https://img.shields.io/crates/v/panopticon-tui.svg)

Terminal UI observability and diagnostic tool for Scala applications.

Currently supports:
- [ZIO-ZMX](https://github.com/zio/zio-zmx)
- [Slick + HikariCP](https://scala-slick.org/doc/3.2.0/config.html#monitoring) (over JMX)
- Akka actor metrics (via [akka-periscope](https://github.com/ScalaConsultants/akka-periscope))

## Usage

### Prerequisites

You'll need to have java installed (it is loaded dynamically by [jmx](https://docs.rs/jmx/0.2.1/jmx/) crate).

### Example

There's a complete example of adjusting a real project to work with panopticon. [Check it out](https://github.com/ScalaConsultants/panopticon-example) to see how it's done.

### Running Panopticon

To get a binary for your OS check [releases](https://github.com/ScalaConsultants/panopticon-tui/releases) page.

Panopticon is published to [crates.id](https://crates.io/), so if you have rust toolchain installed, you can run `cargo install panopticon-tui` and it will build the executable for you.

You can also build it from sources:
```
cargo build
./target/debug/panopticon-tui [OPTIONS]
```

To get a detailed help message, run:
```
panopticon-tui --help
```

### ⚠️ MacOS and libjvm.dylib

On MacOS you can face an error like this:
```
dyld: Library not loaded: @rpath/libjvm.dylib
  Referenced from: panopticon-tui
  Reason: image not found
```

It means that [j4rs](https://docs.rs/j4rs), which is used for JMX integration wasn't able to locate your java installation.

To fix that, just create a symlink to `libjvm.dylib` (this works for JDK 11 and MacOS 10.15.4):

```
sudo ln -s $(/usr/libexec/java_home)/lib/server/libjvm.dylib /usr/local/lib
```

Depending on your MacOS version or java package, location may differ, so make sure to check the symlink is valid.


### Connecting to zio-zmx server

[ZIO-ZMX](https://github.com/zio/zio-zmx) is a tool for monitoring ZIO-based apps. With panopticon you can see the fiber tree visualized and monitor their number:

![ZIO tab demo](./assets/zio-demo.png)

You can specify zio-zmx server address with `--zio-zmx` option. In this case Panopticon will connect to it and show a ZMX tab:
```
panopticon-tui --zio-zmx localhost:6789
```

**⚠️ WARNING**: Currently, zio-zmx doesn't provide efficient ways of getting fiber count metrics, so Panopticon has to do a full fiber dump each tick to calculate them. Make sure your `tick-rate` isn't too frequent.

### Database metrics over JMX

Panopticon can show database metrics, if your app exposes them via JMX. Slick and HikariCP are the only supported options at the moment.

![Slick tab demo](./assets/slick-demo.png)

Slick tab will be shown in Panopticon if you specify these two options:

```
panopticon-tui --jmx localhost:9010 --db-pool-name myDb
```

Here `db-pool-name` is a connection pool name, used to qualify JMX beans for Slick and/or HikariCP. 

See [this section](https://scala-slick.org/doc/3.2.0/config.html#monitoring) of Slick docs for details about setting up your app to expose db metrics over JMX.

Also keep in mind that you need to specify some JVM parameters on startup so that your app exposes JMX metrics. Something along the following lines:

```
-Dcom.sun.management.jmxremote
-Dcom.sun.management.jmxremote.port=9010
-Dcom.sun.management.jmxremote.rmi.port=9010
-Dcom.sun.management.jmxremote.local.only=false
-Dcom.sun.management.jmxremote.authenticate=false
-Dcom.sun.management.jmxremote.ssl=false
```

### Akka metrics

Panopticon can also do several valuable things for akka-based systems:
* show entire tree of actors under some actor system;
* monitor total amount of actors in real-time;
* show latest messages that went to "dead-letters" queue with details;
* monitor total amount of dead letters in real-time.


![Akka tab demo](./assets/akka-demo.png)

To use these features, however, you'd have to enable publication of this data in your application. There's the [akka-periscope](https://github.com/ScalaConsultants/akka-periscope) library, specifically suited for that purpose. Checkout it's README for the details on how you can set it up.

Only HTTP way of transfer is supported for at the moment. To use it and see the actor data on a separate tab, launch Panopticon with following options:

```
panopticon-tui --actor-tree http://localhost:8080/actor-tree --actor-system-status http://localhost:8080/actor-system-status --dead-letters http://localhost:8080/dead-letters
```

Replace the endpoint urls with the ones you set up with [akka-periscope](https://github.com/ScalaConsultants/akka-periscope).

## Build from sources

Development build:
```
cargo build
```

Optimized release build:
```
cargo build --release
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [Unreleased]

## [0.2.1]
### Changed
Bumped jmx to 0.2.1 which should fix the issue with auto-locating jdk on MacOS

## [0.2.0]
### Added
- Akka deadletters metrics (via [akka-periscope](https://github.com/ScalaConsultants/akka-periscope))
- Akka actor system start time and uptime (via [akka-periscope](https://github.com/ScalaConsultants/akka-periscope))
### Changed
- `--actor-count` flag renamed into `--actor-system-status`.

## [0.1.1] - 2020-05-25
### Fixed
- Now panopticon will work with a single Akka tab

## [0.1.0] - 2020-05-19
### Added
- ZIO-ZMX integration
- Slick and HikariCP metrics
- Akka metrics integration (via [akka-periscope](https://github.com/ScalaConsultants/akka-periscope))


src/main.rs
mod ui;
mod zio;
mod jmx;
mod akka;
mod app;
mod fetcher;
mod widgets;

use std::{
    env,
    io::{stdout, Write},
    sync::mpsc,
    thread,
    time::{Duration, Instant},
};

use crossterm::{
    event::{self, Event as CEvent, KeyCode},
    execute,
    terminal::{disable_raw_mode, enable_raw_mode, EnterAlternateScreen, LeaveAlternateScreen},
};
use structopt::StructOpt;
use tui::{
    backend::CrosstermBackend,
    Terminal,
};

use crate::app::{App, AppTabKind};
use crate::fetcher::{Fetcher, FetcherRequest, FetcherResponse};

use crate::akka::model::AkkaSettings;
use crate::jmx::model::JMXConnectionSettings;

enum Event<I> {
    Input(I),
    Tick,
    FetcherResponse(FetcherResponse),
}

/// At least one of the following option sets has to be specified for panopticon-tui to launch:
///
/// - zio-zmx
///
/// - jmx + db-pool-name
///
/// - actor-tree + actor-system-status + dead-letters
#[derive(Debug, StructOpt)]
struct Cli {
    /// Frequency (in ms) to use for fetching metrics.
    /// Don't set this too low, because currently zmx tab does a full fiber dump every tick
    #[structopt(long = "tick-rate", default_value = "2000")]
    tick_rate: u64,
    /// Address of zio-zmx server, e.g. localhost:6789
    #[structopt(long = "zio-zmx")]
    zio_zmx: Option<String>,
    /// Address of remote jmx source, e.g. localhost:9010
    #[structopt(long = "jmx")]
    jmx: Option<String>,
    /// Optional username for authorized jmx access
    #[structopt(long = "jmx-username")]
    jmx_username: Option<String>,
    /// Optional password for authorized jmx access
    #[structopt(long = "jmx-password")]
    jmx_password: Option<String>,
    /// Connection pool name, used to qualify JMX beans for Slick and/or HikariCP
    #[structopt(long = "db-pool-name")]
    db_pool_name: Option<String>,
    /// Address of http endpoint to get akka actor tree
    #[structopt(long = "actor-tree")]
    actor_tree: Option<String>,
    /// Address of http endpoint to get current actor system status
    #[structopt(long = "actor-system-status")]
    actor_system_status: Option<String>,
    /// Time period (in ms) to assemble akka actor tree
    #[structopt(long = "actor-tree-timeout", default_value = "1000")]
    actor_tree_timeout: u64,
    /// Address of http endpoint to get akka dead-letters metrics
    #[structopt(long = "dead-letters")]
    dead_letters: Option<String>,
    /// Time window for akka dead-letters metrics
    #[structopt(long = "dead-letters-window", default_value = "5000")]
    dead_letters_window: u64,
    /// if true then panopticon only checks all the requested connections and exits
    #[structopt(long = "check")]
    check: bool,
}

impl Cli {
    fn jmx_settings(&self) -> Option<JMXConnectionSettings> {
        match (&self.jmx, &self.db_pool_name) {
            (Some(addr), Some(db_pool)) => Some(JMXConnectionSettings {
                address: addr.clone(),
                username: self.jmx_username.clone(),
                password: self.jmx_password.clone(),
                db_pool_name: db_pool.clone(),
            }),
            _ => None
        }
    }

    fn akka_settings(&self) -> Option<AkkaSettings> {
        match (&self.actor_tree, &self.actor_system_status, &self.dead_letters) {
            (Some(tree_addr), Some(status_addr), Some(dead_letters)) => Some(AkkaSettings {
                tree_address: tree_addr.to_owned(),
                tree_timeout: self.actor_tree_timeout,
                status_address: status_addr.to_owned(),
                status_timeout: (self.tick_rate as f64 * 0.8) as u64,
                dead_letters_address: dead_letters.to_owned(),
                dead_letters_window: self.dead_letters_window,
            }),
            _ => None
        }
    }
}

fn main() -> Result<(), failure::Error> {
    let cli = Cli::from_args();

    // disable jmx crate logging
    env::set_var("J4RS_CONSOLE_LOG_LEVEL", "disabled");

    if cli.zio_zmx.is_none() && cli.jmx_settings().is_none() && cli.akka_settings().is_none() {
        let mut clap = Cli::clap();
        println!("Nothing to monitor. Please check the following help message.\n");
        clap.print_long_help().expect("Failed printing help message");
        return Ok(());
    }

    let tick_rate = Duration::from_millis(cli.tick_rate);
    let has_jmx = cli.jmx_settings().is_some();
    let has_akka = cli.akka_settings().is_some();
    let has_zio = cli.zio_zmx.is_some();

    let mut stdout = stdout();
    enable_raw_mode()?;

    execute!(stdout, EnterAlternateScreen)?;

    let backend = CrosstermBackend::new(stdout);

    let mut terminal = Terminal::new(backend)?;
    terminal.hide_cursor()?;

    let mut app = App::new(
        "PANOPTICON-TUI",
        cli.zio_zmx.clone(),
        cli.jmx_settings(),
        cli.akka_settings(),
        cli.check,
    );

    terminal.clear()?;

    // channel for main app event loop
    let (tx, rx) = mpsc::channel();

    // Setup fetcher interaction
    let (txf, rxf) = mpsc::channel();
    {
        let tx = tx.clone();
        thread::spawn(move || {
            let respond = |r| tx.send(Event::FetcherResponse(r)).unwrap();

            match Fetcher::new(cli.zio_zmx.clone(),
                               cli.jmx_settings(),
                               cli.akka_settings()) {
                Err(e) => {
                    eprintln!("Responding with failure {}", e);
                    loop {
                        rxf.recv().unwrap();
                        respond(FetcherResponse::FatalFailure(e.to_owned()))
                    }
                }
                Ok(fetcher) =>
                    loop {
                        match rxf.recv().unwrap() {
                            FetcherRequest::FiberDump =>
                                respond(FetcherResponse::FiberDump(fetcher.dump_fibers())),
                            FetcherRequest::RegularFiberDump =>
                                respond(FetcherResponse::RegularFiberDump(fetcher.dump_fibers())),
                            FetcherRequest::HikariMetrics =>
                                respond(FetcherResponse::HikariMetrics(fetcher.get_hikari_metrics())),
                            FetcherRequest::SlickMetrics =>
                                respond(FetcherResponse::SlickMetrics(fetcher.get_slick_metrics())),
                            FetcherRequest::SlickConfig =>
                                respond(FetcherResponse::SlickConfig(fetcher.get_slick_config())),
                            FetcherRequest::ActorTree =>
                                respond(FetcherResponse::ActorTree(fetcher.get_actor_tree())),
                            FetcherRequest::ActorSystemStatus =>
                                respond(FetcherResponse::ActorSystemStatus(fetcher.get_actor_system_status())),
                            FetcherRequest::DeadLetters =>
                                respond(FetcherResponse::DeadLetters(fetcher.get_dead_letters())),
                        }
                    }
            }
        });
    }

    // Setup input handling
    {
        let tx = tx.clone();
        let txf = txf.clone();
        thread::spawn(move || {
            let mut last_tick = Instant::now();

            // prepopulate some manually loaded data
            if has_jmx {
                txf.send(FetcherRequest::SlickConfig).unwrap();
                txf.send(FetcherRequest::HikariMetrics).unwrap();
                txf.send(FetcherRequest::SlickMetrics).unwrap();
            }
            if has_akka {
                txf.send(FetcherRequest::ActorTree).unwrap();
            }
            if has_zio {
                txf.send(FetcherRequest::FiberDump).unwrap();
            }

            loop {
                // poll for tick rate duration, if no events, sent tick event.
                if event::poll(tick_rate - last_tick.elapsed()).unwrap() {
                    if let CEvent::Key(key) = event::read().unwrap() {
                        tx.send(Event::Input(key)).unwrap();
                    }
                }
                if last_tick.elapsed() >= tick_rate {
                    tx.send(Event::Tick).unwrap();
                    last_tick = Instant::now();
                }
            }
        });
    }

    // stub options that are turned-off
    if app.check_status.is_check_mode {
        if !has_akka {
            app.check_status.checks_passed.insert(FetcherRequest::ActorTree);
            app.check_status.checks_passed.insert(FetcherRequest::ActorSystemStatus);
            app.check_status.checks_passed.insert(FetcherRequest::DeadLetters);
        }

        if !has_zio {
            app.check_status.checks_passed.insert(FetcherRequest::FiberDump);
            app.check_status.checks_passed.insert(FetcherRequest::RegularFiberDump);
        }

        if !has_jmx {
            app.check_status.checks_passed.insert(FetcherRequest::SlickConfig);
            app.check_status.checks_passed.insert(FetcherRequest::HikariMetrics);
            app.check_status.checks_passed.insert(FetcherRequest::SlickMetrics);
        }
    }

    loop {
        ui::draw(&mut terminal, &mut app)?;
        match rx.recv()? {
            Event::Input(event) => match event.code {
                KeyCode::Char('q') => {
                    disable_raw_mode()?;
                    execute!(terminal.backend_mut(), LeaveAlternateScreen)?;
                    terminal.show_cursor()?;
                    break;
                }
                KeyCode::Char(c) => app.on_key(c),
                KeyCode::Left => app.on_left(),
                KeyCode::Up => app.on_up(),
                KeyCode::Right => app.on_right(),
                KeyCode::Down => app.on_down(),
                KeyCode::PageUp => app.on_page_up(),
                KeyCode::PageDown => app.on_page_down(),
                KeyCode::Enter => {
                    match app.tabs.current().kind {
                        AppTabKind::ZMX => txf.send(FetcherRequest::FiberDump)?,
                        AppTabKind::Slick => {}
                        AppTabKind::Akka => txf.send(FetcherRequest::ActorTree)?,
                    }
                }
                _ => {}
            },
            Event::FetcherResponse(r) => match r {
                FetcherResponse::FatalFailure(e) =>
                    app.quit(Some(e)),

                FetcherResponse::FiberDump(d) =>
                    match d {
                        Err(e) => app.quit(Some(e)),
                        Ok(x) => {
                            if app.check_status.is_check_mode {
                                app.check_status.checks_passed.insert(FetcherRequest::FiberDump);
                            }
                            app.zmx.as_mut().unwrap().replace_fiber_dump(x);
                        }
                    },
                FetcherResponse::RegularFiberDump(d) =>
                    match d {
                        Err(e) => app.quit(Some(e)),
                        Ok(x) => {
                            if app.check_status.is_check_mode {
                                app.check_status.checks_passed.insert(FetcherRequest::RegularFiberDump);
                            }
                            app.zmx.as_mut().unwrap().append_fiber_dump_for_counts(x)
                        }
                    },
                FetcherResponse::HikariMetrics(d) =>
                    match d {
                        Err(_) => app.slick.as_mut().unwrap().has_hikari = false,
                        Ok(x) => {
                            if app.check_status.is_check_mode {
                                app.check_status.checks_passed.insert(FetcherRequest::HikariMetrics);
                            }
                            app.slick.as_mut().unwrap().has_hikari = true;
                            app.slick.as_mut().unwrap().append_hikari_metrics(x)
                        }
                    },
                FetcherResponse::SlickMetrics(d) =>
                    match d {
                        Err(e) => app.quit(Some(e)),
                        Ok(x) => {
                            if app.check_status.is_check_mode {
                                app.check_status.checks_passed.insert(FetcherRequest::SlickMetrics);
                            }
                            app.slick.as_mut().unwrap().append_slick_metrics(x)
                        }
                    },
                FetcherResponse::SlickConfig(d) =>
                    match d {
                        Err(e) => app.quit(Some(e)),
                        Ok(x) => {
                            if app.check_status.is_check_mode {
                                app.check_status.checks_passed.insert(FetcherRequest::SlickConfig);
                            }
                            app.slick.as_mut().unwrap().replace_slick_config(x)
                        }
                    },
                FetcherResponse::ActorTree(d) =>
                    match d {
                        Err(e) => app.quit(Some(e)),
                        Ok(x) => {
                            if app.check_status.is_check_mode {
                                app.check_status.checks_passed.insert(FetcherRequest::ActorTree);
                            }
                            let akka = app.akka.as_mut().unwrap();
                            akka.update_actor_tree(x);
                            akka.reload_dead_letters_log();
                        }
                    },
                FetcherResponse::ActorSystemStatus(d) =>
                    match d {
                        Err(e) => app.quit(Some(e)),
                        Ok(x) => {
                            if app.check_status.is_check_mode {
                                app.check_status.checks_passed.insert(FetcherRequest::ActorSystemStatus);
                            }
                            app.akka.as_mut().unwrap().append_system_status(x)
                        }
                    },
                FetcherResponse::DeadLetters(d) =>
                    match d {
                        Err(e) => app.quit(Some(e)),
                        Ok(x) => {
                            if app.check_status.is_check_mode {
                                app.check_status.checks_passed.insert(FetcherRequest::DeadLetters);
                            }
                            app.akka.as_mut().unwrap().append_dead_letters(x.0, x.1)
                        }
                    },
            }

            Event::Tick => {
                if app.check_status.is_check_mode {
                    if app.check_status.is_success() {
                        app.quit(None)
                    } else if app.check_status.ticks_passed >= 10 {
                        app.quit(Some("Connection check took too long to pass".to_owned()))
                    } else {
                        app.check_status.ticks_passed += 1;
                    }
                }

                if app.zmx.is_some() {
                    txf.send(FetcherRequest::RegularFiberDump)?;
                }

                match &app.slick {
                    Some(s) => {
                        txf.send(FetcherRequest::SlickMetrics)?;
                        if s.has_hikari {
                            txf.send(FetcherRequest::HikariMetrics)?;
                        }
                    }
                    None => {}
                }

                if app.akka.is_some() {
                    txf.send(FetcherRequest::ActorSystemStatus)?;
                }
                txf.send(FetcherRequest::DeadLetters)?;
            }
        }
        if app.should_quit {
            break;
        }
    }
    execute!(terminal.backend_mut(), LeaveAlternateScreen)?;

    match app.exit_reason {
        None => Ok(()),
        Some(e) => {
            println!("{}", e);
            std::process::exit(1);
        }
    }
}
